Add allocator and datapath plumbing for cookie-based policy logging

tklauser opened the pull request:
This adds the necessary plumbing in the policy engine and the datapath to make the  arbitrary plain-text Log policy rule field introduced in https://github.com/cilium/cilium/pull/39902 useful. This works as follows:

1. A unique 32-bit value (called a cookie) is allocated for each unique policy log string when distilling policy rules. A cookie value of 0 means no cookie (default case). Cookies are managed by a data structure called the cookie bakery (kudos for the name to @squeed).
2. That cookie is written to the endpoint's policy map as part of the policy rule and passed up by the datapath as part of the policy verdict notification message whenever an allow or deny based on the respective rule entry occurs.
3. The cookie (if non-zero) is used to look up the corresponding policy log string from the cookie bakery and (if found) used to populate the respective Hubble flow with the log string.
4. The cookie bakery is garbage collected at a regular interval using a mark-and-sweep approach.

Note that this PR only adds the cookie allocator for step 1 and the necessary datapath plumbing for step 2 (also see https://github.com/cilium/cilium/pull/40023#issuecomment-3389586917). Follow-up changes will wire this up in the policy engine and Hubble policy correlation.

See individual commits for details.

christarazi commented on pkg/policy/logcookie/bitset.go:
```diff
@@ -0,0 +1,87 @@
+// Allocates a bit at the next free offset, starting at 0 and filling any gaps. It also returns
+// whether a bit could be allocated.
+//
+// By allocating sequentially and filling gaps we safe memory compared to a random strategy which in
```
Nit: `save` instead of `safe`

tklauser replied:
Fixed, thanks.

christarazi commented on pkg/policy/selectorcache.go:
```diff
@@ -242,9 +246,10 @@ func (sc *SelectorCache) queueNotifiedUsersCommit(txn *versioned.Tx, wg *sync.Wa
+		logger:     logger,
+		idCache:    make(map[identity.NumericIdentity]scIdentity, len(ids)),
+		selectors:  make(map[string]*identitySelector),
+		logCookies: logcookie.NewBakery[uint32, string](logger),
```
Non-blocking comment: From an architecture point of view, is it necessary for the selectorcache to handle log cookies? For practical reason currently, it's not the worst thing, but it doesn't feel strictly necessary for the selectorcache to take up this new responsibility. It feels like the policy repo is a more natural place for log cookies.

tklauser replied:
I agree that the policy repo is the more natural place for the cookie allocator. Initially, the cookie allocator was part of the policy repo but it was awkward to pass it to functions allocating the cookie (`(*L4Filter).toMapState` and `(*L4Policy) AccumulateMapChanges`). Both functions don't have the policy repo easily available currently. So this is for practical reasons currently. I'll address this in a follow-up refactoring change and add a TODO comment for now.

christarazi reviewed (approved):
Nice work! Non-blocking comments below.

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -1541,6 +1541,15 @@ func (e *Endpoint) syncPolicyMapWithDump() error {
+	// Garbage collect policy log cookies based on policy map dump. Do this after policy map
+	// reconciliation above so currentMap reflects the current state as close as possible.
+	// Avoid iterating the map again in case there are no policy log cookies allocated.
+	if e.policyRepo.LogCookieCount() > 0 {
```
I wonder if we can do this entirely from the PolicyCache? The PolicyCache has every SelectorPolicy, and that has references to every active MapState.

tklauser replied:
Yeah, that would certainly be cleaner. Let me try that...

squeed commented on pkg/policy/mapstate.go:
```diff
@@ -400,17 +400,24 @@ type mapStateEntry struct {
 
 // newMapStateEntry creates a map state entry.
-func newMapStateEntry(derivedFrom ruleOrigin, proxyPort uint16, priority ListenerPriority, deny bool, authReq AuthRequirement) mapStateEntry {
+func newMapStateEntry(
```
I wonder if we can get rid of this constructor entirely? Changing all the call sites when a new optional parameter is added is frustrating.

tklauser replied:
Agree, done in an additional commit.

squeed commented on pkg/policy/rule.go:
```diff
@@ -39,7 +40,12 @@ type rule struct {
+	logger *slog.Logger,
+	selector policytypes.CachedSelector,
+	added, deleted []identity.NumericIdentity,
+	bakery logcookie.PolicyLogBakery,
```
Hmm. What if cookie allocation was done in the MapState?

I'm not sure if this is right, but it's something to explore. And feel free to say so if you already tried this and it was uglier.

tklauser replied:
I think I've tried this but I can't remember why it didn't work out. Let me meditate on this a bit, given there's no hurry with this PR. Together with https://github.com/cilium/cilium/pull/40023#discussion_r2166715425 we should hopefully also be able to let the policy repo own the bakery to address https://github.com/cilium/cilium/pull/40023#discussion_r2153069771

tklauser replied:
Yeah, turns out this becomes slightly uglier and it would also need to live in `newMapStateEntry` which we want to get rid of per https://github.com/cilium/cilium/pull/40023#discussion_r2166730322. So I think I'll leave this as is for now and probably refactor as a follow-up.

thorn3r reviewed (approved):
this is very cool

tklauser committed:
policy: fix godoc comments

(*EndpointPolicy).GetRuleMeta returns a RuleMeta type since commit
5f8f8e0123f8 ("hubble: add Log to hubble rule metadata"), so the comment
about the return type is no longer accurate.

Also fix the casing for the {selector,policy}Cache members of type
Repository.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

kaworu commented on pkg/hubble/parser/threefour/parser.go:
```diff
@@ -280,7 +287,29 @@ func (p *Parser) Decode(data []byte, decoded *pb.Flow) error {
+			}
+		}
+	}
+
+	if !correlated && p.correlateL3L4Policy && p.endpointGetter != nil {
 		correlation.CorrelatePolicy(p.log, p.endpointGetter, decoded)
 	}
```
LGTM but nitpicking: can we extract that logic into a func dedicated to policy correlation so we avoid adding more code to the already-too-long Decode func please? :pray:

tklauser committed:
policy/cookie: add cookie allocator

Bakery is a cookie allocator, allocating unique cookies of unsigned
integer type for values of comparable type (commonly a string). It
allows looking up the value for a given cookie and mark-and-sweep
garbage collection to remove unused cookies. The implementation is by a
bitset. The bitset is currently implemented on top of a big.Int for
simplicity, but in case memory consumption becomes an issue it could
easily be swapped out for a compressing bitset, e.g. Roaring Bitmap[^1].

[^1]: http://roaringbitmap.org

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
bpf, maps, monitor, policy: wire up policy log cookie

Wire the policy log cookie through the datapath. It is added to the
policy map value (struct policy_entry) and reported back to user space
in the policy verdict message (struct policy_verdict_notify).

For now the cookie is always zeroed. A successive commit will set the
coookie in the agent upon policy rule insertion (if a policy log string
is present in the corresponding policy) and use it in Hubble policy
correlation to look up the corresponding policy log string on policy
verdict notifications.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
cilium-dbg: support policy log cookie

Report the policy log cookie in the output of the `cilium-dbg bpf
policy list` command and allow to set the cookie when adding a new
policy entry using `cilium-dbg bpf policy add`.

Signed-off-by: Tobias Klauser <tobias@cilium.io>
